crates/rune-cli/src/tests.rs

```diff
@@ -0,0 +1,131 @@
+    mut out: StandardStream,
+    runtime: Arc<RuntimeContext>,
+    unit: Arc<Unit>,
+    sources: Sources,
+    tests: Vec<(Hash, CompileMeta)>,
+) -> anyhow::Result<ExitCode> {
+    // TODO: use rune-tests capture_output to stop prints from tests from showing
+    let start = Instant::now();
+    let mut failures = HashMap::new();
+
+    for test in &tests {
+        write!(out, "testing {:40} ", test.1.item.item)?;
+        let mut vm = Vm::new(runtime.clone(), unit.clone());
+
+        let info = unit.lookup(test.0).ok_or_else(|| {
+            VmError::from(VmErrorKind::MissingEntry {
+                hash: test.0,
+                item: test.1.item.item.clone(),
+            })
+        })?;
+
+        let offset = match info {
+            // NB: we ignore the calling convention.
+            // everything is just async when called externally.
+            UnitFn::Offset { offset, .. } => offset,
+            _ => {
+                return Err(VmError::from(VmErrorKind::MissingFunction { hash: test.0 }).into());
+            }
+        };
+
+        vm.set_ip(offset);
+        match vm.async_complete().await {
+            Err(e) => {
+                // TODO: store output here
+                failures.insert(test.1.item.item.clone(), FailureReason::Crash(e));
+                writeln!(out, "crashed")?;
+            }
+            Ok(v) => {
+                if let Ok(v) = v.clone().into_result() {
+                    let res = v.take().unwrap();
+                    if res.is_err() {
+                        failures.insert(test.1.item.item.clone(), FailureReason::ReturnedErr(res));
+                        writeln!(out, "returned error")?;
+                    }
+
+                    continue;
+                }
+                if let Ok(v) = v.into_option() {
+                    if v.borrow_ref().unwrap().is_none() {
+                        failures.insert(test.1.item.item.clone(), FailureReason::ReturnedNone);
+                        writeln!(out, "returned none")?;
+                    }
+
+                    continue;
+                }
+
+                writeln!(out, "passed")?;
+            }
+        }
+    }
+
+    let elapsed = start.elapsed();
+
+    let failure_count = failures.len();
+    for (item, error) in failures {
+        let mut writer = StandardStream::stderr(ColorChoice::Always);
+        match error {
+            FailureReason::Crash(err) => {
+                println!("----------------------------------------");
+                println!("Test: {}\n", item);
+                err.emit_diagnostics(&mut writer, &sources)
+                    .expect("failed writing info");
+            }
+            FailureReason::ReturnedNone => continue,
+            FailureReason::ReturnedErr(e) => {
+                println!("----------------------------------------");
+                println!("Test: {}\n", item);
+                println!("Return value: {:?}\n", e);
+            }
+        }
+    }
```

Comment:
I think moving all of this information into a test-case struct would make this clearer.